src/deployer.ts

```diff
@@ -0,0 +1,194 @@
+import {
+  ChainMap,
+  ChainName,
+  CoreContracts,
+  HyperlaneAddresses,
+  HyperlaneContracts,
+  HyperlaneCoreDeployer,
+  HyperlaneIgpDeployer,
+  MultiProvider,
+  objMap,
+  objMerge,
+  serializeContracts,
+} from '@hyperlane-xyz/sdk';
+import yargs from 'yargs';
+
+import { LegacyMultisigIsm } from '@hyperlane-xyz/core';
+import { ethers } from 'ethers';
+import { multisigIsmConfig } from '../config/multisig_ism';
+import { startBlocks } from '../config/start_blocks';
+import {
+  assertBalances,
+  assertBytes32,
+  buildCoreConfig,
+  buildIgpConfig,
+  buildOverriddenAgentConfig,
+  getMultiProvider,
+} from './config';
+import { mergeJSON, writeJSON } from './json';
+import {
+  HyperlaneTestRecipientDeployer,
+  TestRecipientConfig,
+} from './TestRecipientDeployer';
+
+// HyperlanePermissionlessDeployer
+//   Has HyperlaneCoreDeployer
+//   Has HyperlaneIgpDeployer
+//   Has HyperlaneTestRecipientDeployer
+// Loads configs for each from yargs
+//   CoreConfig -> Signer is owner + MultisigIsm from configs
+//   OverheadIgpConfig -> ???
+//   TestRecipientConfig -> ???
+//
+// 0. Configure your local chain
+//      Output: Add ChainMetadata to chains.ts
+// 1. Generate your validator addresses
+//      Output: Add MultisigIsmConfig to multisig_ism.ts
+// 2. Deploy all contracts
+//      Output: Contract addresses in addresses.json, agent config in agent_config.json
+// 3. Run validators
+// 4. Run relayers
+// 5. Send test messages
+
+export function getArgs(multiProvider: MultiProvider) {
+  // For each chain, we need:
+  //   - ChainMetadata for the MultiProvider
+  //   - A MultisigIsmConfig
+  const { intersection } = multiProvider.intersect(
+    Object.keys(multisigIsmConfig),
+  );
+
+  return yargs(process.argv.slice(2))
+    .describe('local', 'The chain to deploy to')
+    .choices('local', intersection)
+    .demandOption('local')
+    .array('remotes')
+    .describe(
+      'remotes',
+      "The chains with which 'local' will be able to send and receive messages",
+    )
+    .choices('remotes', intersection)
+    .demandOption('remotes')
+    .describe('key', 'A hexadecimal private key for transaction signing')
+    .string('key')
+    .coerce('key', assertBytes32)
+    .demandOption('key')
+    .middleware(
+      assertBalances(multiProvider, (argv) => argv.remotes.concat(argv.local)),
+    )
+    .describe('write-agent-config', 'Whether or not to write agent config')
+    .default('write-agent-config', true)
+    .boolean('write-agent-config').argv;
+}
+
+type MultisigIsmContracts = {
+  multisigIsm: LegacyMultisigIsm;
+};
+
+export class HyperlanePermissionlessDeployer {
+  constructor(
+    protected multiProvider: MultiProvider,
```

Comment:
Any reason these can't be public in case they're useful to someone at some point?

Reply:
Nope